sdk/core/System.BinaryData/src/BinaryData.cs

```diff
@@ -0,0 +1,258 @@
+        /// byte array.
+        /// </summary>
+        /// <param name="data">The array to wrap.</param>
+        /// <returns>A wrapper over <paramref name="data"/>.</returns>
+        public static BinaryData FromBytes(byte[] data) =>
+            new BinaryData(data);
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance from a string by converting
+        /// the string to bytes using the UTF-8 encoding.
+        /// </summary>
+        /// <param name="data">The string data.</param>
+        /// <returns>A value representing the UTF-8 encoding of <paramref name="data"/>.</returns>
+        public static BinaryData FromString(string data) =>
+            new BinaryData(data);
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance from the specified stream.
+        /// The stream is not disposed by this method.
+        /// </summary>
+        /// <param name="stream">Stream containing the data.</param>
+        /// <returns>A value representing all of the data remaining in <paramref name="stream"/>.</returns>
+        public static BinaryData FromStream(Stream stream) =>
+#pragma warning disable AZC0102 // Do not use GetAwaiter().GetResult().
+            FromStreamAsync(stream, false).GetAwaiter().GetResult();
+#pragma warning restore AZC0102 // Do not use GetAwaiter().GetResult().
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance from the specified stream.
+        /// The stream is not disposed by this method.
+        /// </summary>
+        /// <param name="stream">Stream containing the data.</param>
+        /// <param name="cancellationToken">A token that may be used to cancel the operation.</param>
+        /// <returns>A value representing all of the data remaining in <paramref name="stream"/>.</returns>
+        public static async Task<BinaryData> FromStreamAsync(
+            Stream stream,
+            CancellationToken cancellationToken = default) =>
+            await FromStreamAsync(stream, true, cancellationToken).ConfigureAwait(false);
+
+        private static async Task<BinaryData> FromStreamAsync(
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken = default)
+        {
+            if (stream == null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+            if (stream.CanSeek && (stream.Length - stream.Position) > int.MaxValue)
+            {
+                throw new ArgumentOutOfRangeException(
+                    nameof(stream),
+                    "Stream length must be less than Int32.MaxValue");
+            }
+
+            MemoryStream? inputMemoryStream = stream as MemoryStream;
+            ArraySegment<byte> buffer = default;
+
+            if (inputMemoryStream?.TryGetBuffer(out buffer) ?? false)
+            {
+                return new BinaryData(buffer.Array.AsMemory(buffer.Offset, buffer.Count));
```

Comment:
Please make sure there's a test like

```C#
byte[] buf = { (byte)0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
MemoryStream memoryStream = new MemoryStream(buf, 1, 5);
memoryStream.ReadByte();
byte[] after = memoryStream.FromStream(memoryStream).WhateverIsNeeded.ToArray();
```

after should be { 2, 3, 4, 5 }.

I think the current code will produce { 1, 2, 3, 4, 5 }, because the returned segment says what buffer it was originally provided, and disregards the Position offset.

Also, FWIW, this line is slightly faster, because it avoids the intermediate representation as a Memory&lt;T>

```suggestion
                return new BinaryData((ReadOnlyMemory<byte>)buffer);
```

Reply:
Good catch - we need to take the stream position into account, but I think we still need to consider the buffer offset as well:
```cs
            if (inputMemoryStream?.TryGetBuffer(out buffer) ?? false)
            {
                return new BinaryData(new ReadOnlyMemory<byte>(buffer.Array, buffer.Offset, buffer.Count).Slice((int)stream.Position, streamLength));
            }
```

Reply:
Actually, I guess TryGetBuffer would only be true when an array buffer is used so it should never have an Offset in this case.

Reply:
Ah the implicit cast handles that for us..